src/relax/transform/extract_dataflow_blocks.cc

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/relax/transform/extract_dataflow_blocks.cc
+ * \brief Pass for extracting groups of pure operations without
+ *   dataflow into dataflow blocks.
+ */
+
+#include <tvm/relax/expr.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/transform.h>
+#include <tvm/relax/utils.h>
+
+namespace tvm {
+namespace relax {
+
+class DataflowBlockExtractor : public ExprMutator {
+ public:
+  explicit DataflowBlockExtractor(size_t min_size) : ExprMutator(), min_size_(min_size) {}
+
+  Expr VisitExpr_(const SeqExprNode* seq) override {
+    Array<BindingBlock> new_blocks;
+    Expr new_body = VisitExpr(seq->body);
+    bool changed = !new_body.same_as(seq->body);
+    for (auto block : seq->blocks) {
+      BindingBlock new_block = this->VisitBindingBlock(block);
+      changed = changed || !new_block.same_as(block);
+      if (new_block.as<DataflowBlock>()) {
```

Comment:
Would be good to comment that we don't need to explicitly merge a new dataflow block with a preceding dataflow block, as that is handled during `IRNormalizer`.

Reply:
Good point. I rely on normalization throughout this pass.